plans/data-transfer-seeds-leeches/utils/metrics.go

```diff
@@ -0,0 +1,14 @@
+package utils
+
+import "github.com/ipfs/testground/sdk/runtime"
+
+var (
+	MetricTimeToFetch = &runtime.MetricDefinition{Name: "time_to_fetch", Unit: "ns", ImprovementDir: -1}
+	MetricMsgsRcvd    = &runtime.MetricDefinition{Name: "msgs_rcvd", Unit: "messages", ImprovementDir: -1}
+	MetricDataSent    = &runtime.MetricDefinition{Name: "data_sent", Unit: "bytes", ImprovementDir: -1}
+	MetricDataRcvd    = &runtime.MetricDefinition{Name: "data_rcvd", Unit: "bytes", ImprovementDir: -1}
+	MetricDupDataRcvd = &runtime.MetricDefinition{Name: "dup_data_rcvd", Unit: "bytes", ImprovementDir: -1}
+	MetricBlksSent    = &runtime.MetricDefinition{Name: "blks_sent", Unit: "blocks", ImprovementDir: -1}
+	MetricBlksRcvd    = &runtime.MetricDefinition{Name: "blks_rcvd", Unit: "blocks", ImprovementDir: -1}
+	MetricDupBlksRcvd = &runtime.MetricDefinition{Name: "dup_blks_rcvd", Unit: "blocks", ImprovementDir: -1}
+)
```

Comment:
Given that you grab all the logs for this with all the nodes. Would the tool that @jimpick created, ipfs-spy, give you what you need to visualize what happened? If not, what else do you need (to inform the dashboarding design)?

Reply:
I think there are a few different scenarios we will want to (eventually) cover:
1. Canary testing
  Can I safely deploy this to prod
2. Benchmarking
  Is my commit faster / slower than a previous commit
3. Debugging
  - In dev
    As I'm developing I want to see what effect my changes are having, and to be able to see output from logs that I've added in temporarily (without having to commit those logs)
  - In production
    In production we will may have issues we need to debug (eg it's taking a long time for blocks to be retrieved).
4. Visualizations
  What happens to a request / response as it passes through the whole system, how do blocks move around etc

My feeling is that we should focus on 1 & 2 for now, and plan for 3 & 4 shortly.

The current set of statistics should provide enough information for 1 & 2.

Dashboarding is probably most useful for 2 and 3: it is useful to find out which subsystem is causing bottlenecks (is it the Bitswap engine, the request queue, the time between sending a want and receiving back a block etc)

Reply:
I should also mention that Bitswap already has [some instrumentation](https://github.com/ipfs/go-bitswap/blob/dcbe1f29c433e1c85705f7239d189e9aed910f96/bitswap.go#L361), but we will probably want to have more to identify bottlenecks in particular subsystems (eg measure the time between a request being pushed and popped from the engine's request queue)